### DISCUSSION TITLE
.populate() and .toJSON() override

### DISCUSSION THREAD

**Discussion by tamis-laan** [Automattic/mongoose#11899] [Q&A]
So I have two schema's A and B where A contains a reference to a document of schema B. Both schema's have there `.toJSON` methods overridden in order to transform and remove certain properties when send to the client. This works for documents with schema A or B in isolation. However when populating a document with schema B where a property refers to a document with schema A, the `.toJSON` logic for the property with schema A is not applied.

How should I fix this?

**Comment by vkarpov15** [comment]
@tamis-laan can you please provide a script that demonstrates the behavior you're seeing?

**Comment by tamis-laan** [comment]
``` javascript
const ASchema = Schema({
  b = {type: ObjectID, ref:'B'}
})

const Bschema = Schema({
  name: String
})

BSchema.method('toJSON', function() {
  let b = this.toObject()
  b.message = `Hello ${b.name}`
  return b
})

const a = await A.findById(someid).populate('b')

const json = a.toJSON()
``` 

The json doesn't include `a.b.message` as toJSON for the B schema is never executed.

**Reply by vkarpov15** [reply]
@tamis-laan I'm unable to repro, below script prints "toJSON called" and adds `a.b.message`. Please modify the below script to demonstrate your issue.

```javascript
const mongoose = require('mongoose');
const { Schema } = mongoose;

void async function main() {
  await mongoose.connect('mongodb://localhost:27017/test');

  const ASchema = Schema({
    b: {type: 'ObjectId', ref:'B'}
  })

  const BSchema = Schema({
    name: String
  })

  BSchema.method('toJSON', function() {
    console.log('toJSON called');
    let b = this.toObject()
    b.message = `Hello ${b.name}`
    return b
  })

  const A = mongoose.model('A', ASchema);
  const B = mongoose.model('B', BSchema);

  const b = await B.create({ name: 'foo' });
  const { _id } = await A.create({ b });

  const a = await A.findById(_id).populate('b')

  const json = a.toJSON()
  console.log(json);
}()
```
